README.markdown
# Setting up Validator

## Things you'll need

	* PHP
	* PHPUnit
	* Frontend Test Suite (git://github.com/NeilCrosby/frontend-test-suite.git)
	* Apache ANT

## Setup

        * git clone git://github.com/robb1e/Validator.git
        * cd Validator
        * git submodule init
        * git submodule update
	
## How to run

To run Validator, edit the build.xml file, replacing the SITE_URL filter token with your desired endpoint.  Then run 'ant validate' from the terminal and the application will start at the page, discover all links collecting those.  It will then run against each of those links reporting pages which return 500 and 404 HTTP status code, and will run HTML validation against those which return successfully.

Ant will then fail if any of these assertions fail.

## Integration

You can integrate this as part of a website build process, or as part of a monitoring solution.  This may be particularly useful for websites with a content management part to protect against those pesky users who might copy and paste from Word.

# Further work

	* CSS Validation
	* Javascript Lint
	* Better reporting


PHPCrawl_070/CHANGELOG.txt
Version 0.7
2007/01/05 

* New function setLinkPriority() added. Its possible now to set different
priorities (priority-levels) for different links.

* Found links are cached now in a different way so that the general
performance was improved, especially when cralwing a lot of pages (big sites)
in one process.

* Added support for Secure Sockets Layer (SSL), so links like "https://.."
will be followed corretly now.
Please note that the PHP OpenSSL-extension is required for it to work.

* The link-extraction and other parts were redone and should give better results
now.

* Methods setAggressiveLinkExtraction() and addLinkExtractionTag() added for
setting up the link extraction manually.

* Added support for robots.txt-files, method obeyRobotsTxt() added.

* More information about found pages and files will be passed now to the method
handlePageData(), especially information about all found links in the current
page is available now. Method disableExtendedLinkInfo() added.

* The crawler now is able to handle links like "http://www.foo.com:123" correctly
and will send requests to the correct port. Method setPort() added.

* The content of pages and files now can be received/streamed diretly to a temporary
file instead of to memory, so now it shouldn't be a problem anymore to let the
crawler receive big files.
Methods setTmpFile(), addReceiveToTmpFileMatch() and addReceiveToMemoryMatch() added.

* Added support for basic authentication. Now the crawler is able to crawl protected
content. Method addBasicAuthentication() added.

* Now its possible to abort the crawling-process by letting the overridable function
handlePageData() return any negative value.

* Method setUserAgentString() added for setting the user-agent-string in
request-headers the crawler sends.

* A html-testinterface for testing the different setups of the crawler is included in
the package now.

* The crawler doesn't do DNS-lookups anymore for every single page-request, only if the
host is changing a lookup will be done once. This improves performance a little bit.

* The crawler doesn't look for links anymore in every file it finds, only
"text/html"-content will be checked for links.

* Fixed problem with rebuilding links like "*foo".

* Fixed problem with splitting/rebuilding URLs like
"http://foo.bar.com/?http://foo/bar.html".

* Fixed problem that the crawler handled i.e. "foo.com" and "www.foo.com" as different
hosts.


Version 0.65_pl2
2005/08/22

* Phpcrawl now doesn't throw any notices anymore when error-reporting
  is set to E_ALL. Thanks to elitec0der!

* Also there shouldn't be any notices anymore when "allow_call_time_pass_reference"
  is set to OFF. Thanks philipx!
  
* Fixed a bug that appeared when running the crawler in followmode 3.
  (The crawler never went "path up" anymore)
  
* Now the crawler sends the referer-header with requests correctly again.


Version 0.65_pl1 (patchlevel 1)
2004/08/06

Just a few bugs were fixed in this version:
(Yes, it took a time, sorry)

* The crawler doesn't send empty "Cookie:" header anymore if there's no cookie to send.
  (A few webserver rerturned a "bad request" header to this) Thanks Mark!

* The crawler doesn't send one and the same cookie-header several times with a request
  anymore if it was set more than one time by the server. (Didn't really matter though)

* Crawler will find links in metatags used with single quotation marks (') like
  <META HTTP-EQUIV='refresh' CONTENT='foo; URL=http://bar/'> now. Thanks Arsa!

* HTTP 1.0 requests will be send now because of problems with HTTP 1.1 headers and
  chunked content. Thanks Arsa again! 

PHPCrawl_070/example.php
<?php

// It may take a whils to crawl a site ...
set_time_limit(10000);

// Inculde the phpcrawl-mainclass
include("classes/phpcrawler.class.php");

// Extend the class and override the handlePageData()-method
class MyCrawler extends PHPCrawler 
{
  function handlePageData(&$page_data) 
  {
    // Here comes your code.
    // Do whatever you want with the information given in the
    // array $page_data about a page or file that the crawler actually found.
    // See a complete list of elements the array will contain in the 
    // class-refenence.
    // This is just a simple example.
    
    // Print the URL of the actual requested page or file
    echo "Page requested: ".$page_data["url"]."<br>";
    
    // Print the first line of the header the server sent (HTTP-status)
    echo "Status: ".strtok($page_data["header"], "\n")."<br>";
    
    // Print the referer
    echo "Referer-page: ".$page_data["referer_url"]."<br>";
    
    // Print if the content was be recieved or not
    if ($page_data["received"]==true)
      echo "Content received: ".$page_data["bytes_received"]." bytes";
    else
      echo "Content not received";
    
    // ...
    
    // Now you should do something with the content of the actual
    // received page or file ($page_data[source]), we skip it in this example
    
    echo "<br><br>";
    echo str_pad(" ", 5000); // "Force flush", workaround
    flush();
  }
}

// Now, create an instance of the class, set the behaviour
// of the crawler (see class-reference for more methods)
// and start the crawling-process.

$crawler = &new MyCrawler();

// URL to crawl
$crawler->setURL("www.php.net");

// Only receive content of files with content-type "text/html"
// (regular expression, preg)
$crawler->addReceiveContentType("/text\/html/");

// Ignore links to pictures, dont even request pictures
// (preg_match)
$crawler->addNonFollowMatch("/.(jpg|gif|png)$/ i");

// Store and send cookie-data like a browser does
$crawler->setCookieHandling(true);

// Set the traffic-limit to 1 MB (in bytes,
// for testing we dont want to "suck" the whole site)
$crawler->setTrafficLimit(1000 * 1024);

// Thats enough, now here we go
$crawler->go();


// At the end, after the process is finished, we print a short
// report (see method getReport() for more information)

$report = $crawler->getReport();

echo "Summary:<br>";
if ($report["traffic_limit_reached"]==true)
  echo "Traffic-limit reached <br>";
  
echo "Links followed: ".$report["links_followed"]."<br>";
echo "Files received: ".$report["files_received"]."<br>";
echo "Bytes received: ".$report["bytes_received"]."<br>";

?>

PHPCrawl_070/test_interface/output.php
<?php

// This file shouldn't be opened/run directly 
if (!preg_match("#index\.php$#i", $_SERVER["SCRIPT_FILENAME"]))
{
  header("Location: index.php");
  exit;
}

?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">

<html>
<head>
	<title>PHPCrawl Testinterface Output</title>
 <link rel="stylesheet" type="text/css" href="style.css">
  
 <style type="text/css">
 #cont {
   height:100%;
 }
 
	#output {
	   width:691px; 
	   height:500px; 
	   background-color:#e1e1e1; 
	   border-width:2px; 
	   border-style:solid; 
	   overflow:auto;
	   font-family: verdana;
	   font-size: 10px;
    padding-left: 5px;
	 }
 
 #head {
    padding-left: 5px;
    padding-top: 3px;
	   width:691px; 
	   height:30px; 
	   background-color:#ffffff; 
	   border-top: 2px black solid;
    border-left: 2px black solid; 
    border-right: 2px black solid; 
	   font-family: verdana;
	   font-size: 14px;
    font-weight: bold;
	 }
  
  #foot {
	   width:691px; 
	   height:40px; 
	   background-color:#ffffff; 
	   border-bottom: 2px black solid;
    border-left: 2px black solid; 
    border-right: 2px black solid; 
	   font-family: verdana;
	   font-size: 12px;
    padding-left: 5px;
    padding-top: 2px;
    vertical-align: bottom;
	 }
  
	 table.intbl
	 {
	   border-color:#000000;
    border-collapse: collapse;
	   border-width:1px;
	   border-style:solid;
	 }
 
	 table.intbl td 
	 {
    background-color: #efefef;
	   padding-left:3px;
    border: 1px black solid;
	 }
  
  table.summary
  {
    background-color: #ffffff;
    font-family: verdana;
	   font-size: 12px;
  }
  
  table.summary td
  {
    padding: 2px;
  }
  
  table.summary td#summary_head
  {
    font-weight: bold;
  }
 
 </style>
 
 <script language="javascript">
 self.focus();
 </script>
</head>

<body>

<?php
// Setup the crawler
$crawler = &new phpcrawlSetup ($val, $output); // $val is the posted setup-array,
                                               // $output the posted output-array
                                                 
if (isset($_POST["misc"]["force_flush"]) && $_POST["misc"]["force_flush"] == "1")
{
  $crawler->force_output_flushing = true;
}

$setup_error = $crawler->setupCrawler();
?>

<div id="container">
		
		<div id="head">
		  PHPCrawl Testinterface Output
    (using version: <?php echo $crawler->class_version; ?>)
		</div>
    
		<div id="output">
  
		<?php  
  flush();
		                                              
		// Start crawling if no error occured during setup
		if ($setup_error==false)
  {
		  echo "<br>";
    
		  $crawler->go();
    
    // Get the report after its finished.
    $summary = $crawler->getReport();
		}
		else
  {
		  echo "<br><font class=warning>Setup error: ".$setup_error."</font>";
		}
		?>
    
		</div>
    
  <div id="foot">
    
    <?php
    // Print the summary if no error occured
    if ($setup_error == false)
    {
		    echo '<table class="summary">
									     <tr>
											     <td id="summary_head">Process finished!&nbsp;</td>
									       <td>Links followed:</td>
									       <td>'.$summary["links_followed"].'&nbsp;&nbsp;</td>
									       <td>Kb received:</td>
									       <td>'.round($summary["bytes_received"] / 1000).'&nbsp;&nbsp;</td>
                <td>Data throughput kb/s:</td>
									       <td>'.round($summary["data_throughput"] / 1000).'&nbsp;&nbsp;</td>
									     </tr>
									     <tr>
									       <td>&nbsp;</td>
									       <td>Files received:</td>
									       <td>'.$summary["files_received"].'&nbsp;&nbsp;</td>
									       <td>Time in sec:</td>
									       <td>'.number_format($summary["process_runtime"], 2, ".", "").'&nbsp;&nbsp;</td>
                <td>&nbsp;</td>
									       <td>&nbsp;</td>
									     </tr>
											  </table>';
    }
    ?>
       
		</div>
    
</div>

</body>
</html>


test.template.php
<?php
// It may take a whils to crawl a site ...
set_time_limit(10000);
// Inculde the phpcrawl-mainclass
// require_once 'PHPUnit/Framework.php';

print_r("\n\n***************\n");
print_r("***************\n");
print_r("There will be a series of warnings about PHPCrawler* classes not loading\n");
print_r("You can safely ignore these warnings, as they are loaded, but frontend-test-suite cannot find them.\n");
print_r("***************\n");
print_r("***************\n\n");

require_once('frontend-test-suite/suite/TheCodeTrainEasyFrontendTestSuite.php' );
include("PHPCrawl_070/classes/phpcrawler.class.php");
// Extend the class and override the handlePageData()-method
class MyCrawler extends PHPCrawler 
{
  public $pages = array();
  public $error_pages = array();
  public $notfound_pages = array();
	
  function handlePageData(&$page_data) 
  {
	if (strstr($page_data["header"], "200 OK"))
    	array_push($this->pages, $page_data["url"]);

	if (strstr($page_data["header"], "Internal Server Error"))
		array_push($this->error_pages, $page_data["url"]." - referrer - ".$page_data["referer_url"]);
	
	if (strstr($page_data["header"], "Not Found"))
		array_push($this->notfound_pages, $page_data["url"]." - referrer - ".$page_data["referer_url"]);

    echo str_pad(".", 1); // "Force flush", workaround
    flush();
  }
}

class ExampleTestSuite extends TheCodeTrainEasyFrontendTestSuite {
    public static function suite() {

		$url="@SITE_URL@";
		$crawler = &new MyCrawler();
		// URL to crawl
		$crawler->setURL($url);
		// Only receive content of files with content-type "text/html"
		// (regular expression, preg)
		$crawler->addReceiveContentType("/text\/html/");
		// Ignore links to pictures, dont even request pictures
		// (preg_match)
		$crawler->addNonFollowMatch("/.(jpg|gif|png|css|js|swf)$/ i");
		// Store and send cookie-data like a browser does
		$crawler->setCookieHandling(true);
		// Set the traffic-limit to 10 MB (in bytes,
		// for testing we dont want to "suck" the whole site)
		$crawler->setTrafficLimit(10000 * 1024);
		// Thats enough, now here we go
		$crawler->go();
		// At the end, after the process is finished, we print a short
		// report (see method getReport() for more information)
		$report = $crawler->getReport();
		echo "Summary:\n";
		if ($report["traffic_limit_reached"]==true)
		  echo "Traffic-limit reached \n";
		print_r($report); 
		echo "\npages\n";
		print_r(array_unique($crawler->pages));
		echo "\nerrors\n\n";
		print_r(array_unique($crawler->error_pages));
		echo "\nnot found\n\n";
		print_r(array_unique($crawler->notfound_pages));
      return parent::suite( array(
        'html'  =>  array(
          'validator' => 'http://validator.w3.org/check',
          'tests'     => $crawler->pages
        )
      ) );
    }
  }


?>